First cut at the core sync infrastructure

mhammond opened the pull request:
Builds on #2810. This needs the trait from Lina's patch, but is setup to work with it. It implements the 3-way merge (although doesn't quite get deletions right yet, so there's still some work to do) - but comments welcome.

thomcc commented:
Regarding how to do the two/three way merge, I'm pretty happy with how https://github.com/mozilla/application-services/blob/master/components/logins/src/update_plan.rs#L34-L60 works, and it's a high level description of the set of operations you perform for a three way merge.

https://github.com/mozilla/application-services/blob/master/components/logins/src/db.rs#L1061-L1114 is the code that calls it, and decides how to handle deletions vs missing mirror, etc. I think trying to handle it as you have is going to be tricky (but your comment says just about as much).

That said, handing things in bulk as this does is clearly not necessary, it could still be done piecemeal.

mhammond commented on components/webext-storage/src/sync/incoming.rs:
```diff
@@ -0,0 +1,479 @@
+    /// We should locally delete the data for this record
+    DeleteLocally,
+    /// We should remotely delete the data for this record
+    DeleteRemotely,
```
incoming should never cause remote deletion

mhammond commented on components/webext-storage/src/sync/incoming.rs:
```diff
@@ -0,0 +1,479 @@
+        IncomingState::LocalOnly { incoming, local } => {
+            // So we have a local record and an incoming/staging record, but *not* a
+            // mirror record. This is the first time we've seen this (ie, almost
+            // certainly another device synced something)
```
another *first* synced something

mhammond commented on components/webext-storage/src/sync/incoming.rs:
```diff
@@ -0,0 +1,479 @@
+    }
+}
+
+pub fn apply_actions<S: ?Sized + Interruptee>(
```
comments

rfk commented on components/webext-storage/src/sync/outgoing.rs:
```diff
@@ -0,0 +1,147 @@
+                ":ext_id": state.ext_id,
+            },
+        )?;
+    }
```
I understand this code isn't all plumbed together yet, but would I be correct to expect that this function should also update the mirror to reflect what was uploaded to the server? Or, will we sync down our own changes and only update the mirror at that point?

mhammond replied:
yes, exactly. That's in a later version too.

rfk commented on components/webext-storage/sql/create_schema.sql:
```diff
@@ -5,19 +5,20 @@
+    CHECK(sync_status = 2 OR data IS NOT NULL)
 ) WITHOUT ROWID;
 
+/* index on sync_status? */
```
That has the potential to be a very churny index, basically every write to this table will also need to update the index. It could be better on balance to pay the full table scan on each sync (not entirely sure how to test the two options in practice tho)

mhammond commented on components/webext-storage/src/sync/outgoing.rs:
```diff
@@ -0,0 +1,147 @@
+        Ok(OutgoingInfo {
+            state: OutgoingStateHolder {
+                ext_id: ext_id.clone(),
+                change_counter: row.get("sync_change_counter")?,
```
yes, exactly that. It could be a bool if the entire sync happened in a single transaction. Instead, we track what the value was when starting, then decrement it by that later. This doesn't have all the change counter stuff wired up (a later version is far more complete)

mhammond commented on components/webext-storage/src/sync/incoming.rs:
```diff
@@ -0,0 +1,479 @@
+                }
+                (Some(_), None) => {
+                    // We've data locally, but there's an incoming deletion.
+                    // Remote wins.
```
Yes - what's on the server always wins. However, it's not clear what the exact semantics are here - see below.

mhammond commented on components/webext-storage/src/sync/mod.rs:
```diff
@@ -0,0 +1,165 @@
+// XXX - this needs more thought, and probably needs significant changes.
+// Main problem is that it doesn't handle deletions - but to do that, we need
+// something other than a simple Option<JsonMap> - we need to differentiate
+// "doesn't exist" from "removed".
```
Yes, we could do this, but TBH I'm not sure what the exact semantics are here - it's not documented, so we probably need to check how chrome behaves in these cases. I've just punted doing anything else here for now as hopefully @glasserc is tackling that as part of #2900

linacambridge replied:
Summarizing the discussion from that thread, it sounds like Chrome's behavior is too coarse, and not really what we want...but also, the Kinto "client wins" strategy isn't right for us, and it's more important to maintain compatibility with what we had before than with Chrome's implementation. Is that accurate?

Since we have the mirror copy, @rfk's suggestion makes sense!

linacambridge commented on components/webext-storage/src/sync/incoming.rs:
```diff
@@ -0,0 +1,470 @@
+                    "UPDATE storage_sync_data SET data = :data, sync_change_counter = sync_change_counter + 1 WHERE ext_id = :ext_id",
+                    &[
+                        (":ext_id", &item.ext_id),
+                        (":data", &serde_json::Value::Object(data).to_string()),
```
Micro-nit: It looks like Rusqlite will [automatically implement `ToSql`](https://github.com/rusqlite/rusqlite/blob/d88fe1c1b18f308b85f38ba0cf89e1dff6998638/src/types/mod.rs#L61-L62) if it's compiled with the `serde_json` feature? Paging @thomcc to make sure! ☎️

mhammond replied:
> Micro-nit: It looks like Rusqlite will [automatically implement `ToSql`](https://github.com/rusqlite/rusqlite/blob/d88fe1c1b18f308b85f38ba0cf89e1dff6998638/src/types/mod.rs#L61-L62) if it's compiled with the `serde_json` feature?

It does indeed - nice catch!

linacambridge commented on components/webext-storage/src/sync/incoming.rs:
```diff
@@ -0,0 +1,470 @@
+
+    // select simple int
+    fn ssi(conn: &Connection, stmt: &str) -> u32 {
+        conn.query_row_and_then(stmt, rusqlite::NO_PARAMS, |row| row.get::<_, u32>(0))
```
I know this is test code, but `try_query_one::<u32>` might replace this, too.

linacambridge commented on components/webext-storage/src/sync/incoming.rs:
```diff
@@ -0,0 +1,470 @@
+        Ok(())
+    }
+
+    // XXX - test apply_actions!
```
Yes, please! 😁 Thanks also for the wonderful tests so far!

linacambridge commented on components/webext-storage/src/sync/mod.rs:
```diff
@@ -0,0 +1,156 @@
+// XXX - need to work out how to consolidate "payload" vs "bso", particularly
+// the timestamp, etc. For now, we represent this in ServerPayload, even though
+// that's never actually stored in this way.
+// XXX - this isn't going to work without other changes.
```
Hmmm, yeah, we're going to need to pass that over in Desktop. I wonder if we can "just" pass the raw decrypted JSON over XPCOM, and have the Desktop integration layer parse it into a `sync15_traits::Payload` (which has `id`, `deleted`, and `data` already on it) before handing it over to `webext_storage`.

linacambridge replied:
I spent a bit of time noodling on this, and the result is #3067 and https://phabricator.services.mozilla.com/D73581. Basically, we have this thing called an "envelope", which is BSO metadata fields + the unparsed payload. An incoming envelope has a `payload()` method, which you can use to get at the payload—which you can then either deserialize into a record, or check if it's a tombstone via `is_deleted`, like you can in a-s `sync15`.

I think we can land your patch as-is now, and, assuming you like the envelope idea, go back and tweak this after.

linacambridge commented on components/webext-storage/src/sync/mod.rs:
```diff
@@ -0,0 +1,156 @@
+            merge(
+                map!({"one": "one", "two": "two"}),
+                map!({"two": "two", "one": "one"}),
+                Some(map!({"parent_only": "parent"})),
```
Hmm, in this case, we still want to update the mirror without reuploading, though, no? (Or maybe it's fine to leave it as is, if it can't cause problems for future merges...say, if local or remote ever brings `parent_only` back).

Or, will we take care of that separately, when we write everything in `storage_sync_staging` back to the mirror?

mhammond replied:
> Or, will we take care of that separately, when we write everything in `storage_sync_staging` back to the mirror?

Yep - the final outgoing process writes staging to the mirror.

linacambridge commented on components/webext-storage/src/sync/outgoing.rs:
```diff
@@ -0,0 +1,178 @@
+            (None, true)
+        };
+        Ok(OutgoingInfo {
+            state: OutgoingStateHolder {
```
Hmmm, this does mean we'll need the Sync JS side to hold on to this, match them up before calling `setUploaded`, and pass them back. That should work also, but how do you feel about storing all these outgoing ones in a temp table, like Places does it? `guid` and `change_counter` would become columns in that table, and when you mark it as uploaded, doing the write back to the mirror?

glasserc commented on components/webext-storage/src/sync/incoming.rs:
```diff
@@ -0,0 +1,470 @@
+/// record. This "state" is the input to calculating the IncomingAction.
+#[derive(Debug, PartialEq)]
+pub enum IncomingState {
+    IncomingOnly {
```
I'd also like to express that the types themselves are confusing to me. I would have assumed that e.g. `IncomingOnly` would require a non-`Option` `JsonMap`. The fact that all of these are `Options` in all variants makes me wonder what the difference is between a simple `struct IncomingState { incoming: Option<JsonMap>, local: Option<JsonMap>, mirror: Option<JsonMap> }`. I'm guessing that the use of `Option` here doesn't necessarily mean "missing" but means something more like tombstones?

mhammond replied:
yes, exactly, it means a tombstone. I can probably make that more explicit by using yet another enum to make it more explicit. I agree the distinction between an empty object and an explicit tombstone isn't clear either.

TBH though, I've been kinda deferring this until this discussion about the semantics.  IIUC, if we went with Chrome's observed behaviour, we wouldn't even need tombstones, but because we are cloning the existing impl, we do, right?

How does the current impl work in this regard? Is there any distinction between an explicit deletion vs an empty object? But as above, I'm not tied to anything here in any way, as I was just ignoring it :)

glasserc replied:
My understanding of Chrome's behavior is that the client has tombstones but the server doesn't. So I think we would still need to represent tombstones to *some* extent. The current implementation does have tombstones both on the client and on the server. Actually, the current implementation has tombstones on a per-key basis, which I don't think we can easily represent given the structures we have now, but maybe I'm missing something.

mhammond replied:
I ended up adding another Enum -

```
/// The state data can be in. Could be represented as Option<JsonMap>, but this
/// is clearer and independent of how the data is stored.
#[derive(Debug, PartialEq)]
pub enum DataState {
    /// The data was deleted.
    Deleted,
    /// Data exists, as stored in the map.
    Exists(JsonMap),
}
```

Which is more verbose, but does make the various states easier to read. I pushed it in rev e478c2142248ad89cf638d0b35d326a45b8101aa - so it's easy to see the changes, but also easy to back out if anyone things it's a backwards step - LMKWYT!

glasserc commented on components/webext-storage/src/sync/mod.rs:
```diff
@@ -0,0 +1,156 @@
+    #[serde(default, skip_serializing_if = "is_default")]
+    data: Option<String>,
+    #[serde(default, skip_serializing_if = "is_default")]
+    deleted: bool,
```
What does it mean if `deleted = false` but `data = Nothing`? Is this the same as an empty object?

mhammond replied:
Yeah, it is representing the actual format on the server. I'm still not 100% happy that's the right choice for the server either though - how is this redundant and potentially ambiguous  option better than an explicit data=null?

mhammond commented on components/webext-storage/src/sync/incoming.rs:
```diff
@@ -0,0 +1,476 @@
+
+/// Details about an incoming item.
+#[derive(Debug, PartialEq)]
+pub struct IncomingItem {
```
I ended up playing with, but not doing this. I was a bit torn as the 3 structs and various tuples does seem a bit clunky at first glance, but:

* `plan_incoming` consumes an `IncomingState` and produces an `IncomingAction` - it doesn't need to know the guid or ext-id. If the `IncomingItem` was part of the state, I'd need to work out how to thread things so the item came out of `plan_incoming()` as part of consuming the state - which would be a bit unnatural as it's not needed there.

* `merge()` makes this even more complicated - it just takes json as inputs and creates an `IncomingAction` as output - which `plan_incoming()` just hands back to its caller - so threading the item all the way through `merge()` makes things quite ugly.

The tl;dr is that in the current setup, all functions use exactly what they need and nothing more. I'm happy to revisit this later though, particularly if Ethan's work makes some of the above no longer true/relevant.

linacambridge replied:
That sounds perfect to me, thanks for taking a look and the thorough response! I agree that the alternatives are even worse, and the way you have this set up now is the cleanest.

mhammond commented:
phew! Thanks the the thorough review @linacambridge! I think I've addressed most review comments here - what's outstanding is:

* Working out what the payload struct should look like - I figured it's best to do that as part of wiring up the sync bridge, and it should be quite easy to retrofit changes here.

* A few outstanding items around merging, for which I'm going to wait for Ethan.

linacambridge commented on components/webext-storage/src/sync/incoming.rs:
```diff
@@ -0,0 +1,648 @@
+/// record. This "state" is the input to calculating the IncomingAction.
+#[derive(Debug, PartialEq)]
+pub enum IncomingState {
+    /// There's an incoming item, but data for that extension doesn't exist
```
I love these comments, thank you so much for adding them! 😍

linacambridge commented on components/webext-storage/src/sync/incoming.rs:
```diff
@@ -0,0 +1,648 @@
+    IncomingOnly { incoming: DataState },
+    /// There's an incoming item and we have data for the same extension in
+    /// our local store - but not in our mirror. This should be relatively
+    /// uncoming as it means:
```
Nit: uncommon

linacambridge commented on components/webext-storage/src/sync/outgoing.rs:
```diff
@@ -0,0 +1,185 @@
+    // table to be done via a single `INSERT INTO ... SELECT` statement - but
+    // due to the GUID complication, and the fact we need to extract each
+    // record from this query anyway means we just do a single query, then
+    // perform the inserts one row at a time.
```
I _think_ you can still do it like...

```sql
INSERT INTO storage_sync_outgoing_staging
(guid, ext_id, data, sync_change_counter)
SELECT coalesce(m.guid, s.guid, generate_guid()),
       l.ext_id, l.data, l.sync_change_counter
FROM storage_sync_data l
LEFT JOIN storage_sync_mirror m ON m.ext_id = l.ext_id
LEFT JOIN storage_sync_staging s ON s.ext_id = l.ext_id
WHERE sync_change_counter > 0
```

...And then read everything out of the temp table, but you're right—since we still need to return a `Vec<...>` of them, anyway, we may as well do it in Rust, and read it from the durable tables before staging! 👍

linacambridge commented on components/webext-storage/src/sync/outgoing.rs:
```diff
@@ -0,0 +1,185 @@
+        items.len()
+    );
+
+    let sql = "
```
I was gonna suggest doing this in a trigger, but we can do it in a follow-up, or leave it be.

linacambridge commented on components/webext-storage/src/sync/outgoing.rs:
```diff
@@ -0,0 +1,185 @@
+        SELECT guid, ext_id, data FROM temp.storage_sync_outgoing_staging;
+
+        -- XXX - is this correct?
+        DELETE from storage_sync_data WHERE data IS NULL AND sync_change_counter = 0;",
```
I think so—durable tombstones get stored in the mirror, so there's no need for them to hang around locally once we've synced them.
